adr/ADR-32.md

```diff
@@ -33,8 +34,8 @@ Service configuration relies on the following:
 - `statsHandler` - an optional function that returns unknown data that can be
   serialized as JSON. The handler will be provided the endpoint for which it is
   building a `EndpointStats`
-- `endpoint` - a subject and a handler (effectively equivalent to a NATS
```

Comment:
This part I don't like - I know we had discussed this between @ripienaar and @Jarema. But here's the issue:

- if you want multiple endpoints you cannot have a handler on the root subject (why not - I hear it is possibly not the right thing, etc but this is an assumption - the app developer will figure out what they want)
- if you want a single endpoint you have add a root subject, and then the named handler, the composite is the subject.

This second thing is what I am more opposed to - the easy case should be here's the subject and the handler.

Reply:
Yeah agree on this, I mentioned earlier I think a single endpoint on the root should be the only one that can specify wildcard subjects and the intention is that it is for just a basic one feature thing (like auth service example I made) or for those making their own request MUX.

If you go multiple entpoints, then you need to follow the convention.

So in some conversation over the holiday period we suggested keeping `endpoint` for the easy path and adding `endpoints` that follow convention - but of course pick another implementation detail that you wish but I agree with @aricart here

Reply:
@ripienaar in retrospect after updating the client and writing some tests, the need for endpoint is not necessarily there, the difference and would simplify the control mechanism - previously it would be added right on the `nc.services.add()`, but a separate call to srv.addEndpoint() achieves the same thing.

```typescript
  const srv = await nc.services.add({
    name: "example",
    version: "0.0.1"
  });
  srv.addEndpoint("add", ()=>{});
  const v1 = srv.addGroup("v1");
  v1.addEndpoint("add", () => {});

  const v2 = srv.addGroup("v2");
  v2.addEndpoint("add", () => {});
  v2.addGroup("internal").addEndpoint(("add"), () => {});
```
Note however perverse that looks it is fairly "legal".

The stats look like this:
@piotrpio  note that the endpoint stats _cannot_ be a map, they have to be a list, as the names for the endpoints could be the same but under different subject. Also note that the subject for each endpoint is included.

```javascript
{
  type: "io.nats.micro.v1.stats_response",
  name: "example",
  id: "NC9LTC35N9Q11ZSZE8BA4Z",
  version: "0.0.1",
  started: "2023-01-06T01:19:19.823Z",
  num_requests: 0,
  num_errors: 0,
  processing_time: 0,
  average_processing_time: 0,
  last_error: undefined,
  endpoints: [
    {
      name: "add",
      subject: "add",
      num_requests: 0,
      num_errors: 0,
      processing_time: 0,
      average_processing_time: 0
    },
    {
      name: "add",
      subject: "v1.add",
      num_requests: 0,
      num_errors: 0,
      processing_time: 0,
      average_processing_time: 0
    },
    {
      name: "add",
      subject: "v2.add",
      num_requests: 0,
      num_errors: 0,
      processing_time: 0,
      average_processing_time: 0
    },
    {
      name: "add",
      subject: "v2.internal.add",
      num_requests: 0,
      num_errors: 0,
      processing_time: 0,
      average_processing_time: 0
    }
  ]
}

```

Reply:
Note above, the metric type information doesn't exist at the top level of the stats, as these are now tracked per endpoint, we could aggregate, but not sure that is important as the raw data is there, and some of the fields like errors couldn't show last_error etc.